Unable to install node module in network guarded by http proxy

**Describe the bug**

Context:

Gitlab CI/CD runners in our organization runs on a limited network and we need to use a web proxy to contact external URL. We pass environment variables to let the system know that it should use http proxy.


Problem: 

When we build a docker image for a matrix bot npm silently fails to download the rust binding necessary for the node module to work.
By explicitly calling the `download-lib.js` script I get the following error that indicates me that the scripts fails to initiate a connection and timeout. This makes me think that the [DownloadHelper](https://github.com/matrix-org/matrix-rust-sdk/blob/471ac07c880b16c8af431ade799e623b1ca5505d/bindings/matrix-sdk-crypto-nodejs/download-lib.js#L24) ignores environment variables that defines which proxy to use.

```
Downloading lib matrix-sdk-crypto.linux-x64-gnu.node from https://github.com/matrix-org/matrix-rust-sdk/releases/download/matrix-sdk-crypto-nodejs-v0.1.0-beta.1/matrix-sdk-crypto.linux-x64-gnu.node
Error: timeout
    at /app/node_modules/node-downloader-helper/dist/index.js:1:13440
    at <anonymous>
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
```


**To Reproduce**
Steps to reproduce the behavior:
1. Have a network proxy that filters HTTP(s) connections
2. Export `http_proxy` and `https_proxy` environment variables with the proxy information
3. Use npm to install modules
4. Rust library is not downloaded and it fails silently

**Expected behavior**
The rust library should be correctly downloaded or at least clearly report that the installation failed during ǹpm install`


**Additional context**
We pass `http_proxy` and `https_proxy` environment variables to our docker build command in the CI to configure applications to use the proxy.

**Logs**

From one test machine in the same network with limited access.

npm install log:
```
bousse-e@tpinfo4:~/tmp/naomod-zoom-bot$ npm i
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated request-promise@4.2.6: request-promise has been deprecated because it extends the now deprecated request package, see https://github.com/request/request/issues/3142
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142

added 171 packages, and audited 172 packages in 43s

22 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npm notice 
npm notice New major version of npm available! 8.5.1 -> 9.1.3
npm notice Changelog: https://github.com/npm/cli/releases/tag/v9.1.3
npm notice Run npm install -g npm@9.1.3 to update!
npm notice 
```

execution of the bot:
```
bousse-e@tpinfo4:~/tmp/naomod-zoom-bot$ node main.js 
/comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js:234
    throw loadError
    ^

Error: Cannot find module '@matrix-org/matrix-sdk-crypto-nodejs-linux-x64-gnu'
Require stack:
- /comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js
- /comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/matrix-bot-sdk/lib/e2ee/CryptoClient.js
- /comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/matrix-bot-sdk/lib/index.js
- /comptes/bousse-e/tmp/naomod-zoom-bot/main.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.Module._load (node:internal/modules/cjs/loader:778:27)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (/comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js:174:31)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Module.require (node:internal/modules/cjs/loader:1005:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/@matrix-org/matrix-sdk-crypto-nodejs/index.js',
    '/comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/matrix-bot-sdk/lib/e2ee/CryptoClient.js',
    '/comptes/bousse-e/tmp/naomod-zoom-bot/node_modules/matrix-bot-sdk/lib/index.js',
    '/comptes/bousse-e/tmp/naomod-zoom-bot/main.js'
  ]
}
```

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -29,8 +29,9 @@ function download_lib(libname) {
 
     var proxy = process.env.https_proxy ?? process.env.http_proxy ?? undefined
     if (proxy) {
+      console.info("Using configured HTTP proxy: " + proxy)
       const proxyAgent = new HttpsProxyAgent(proxy);
-      dl.updateOptions( { agent: proxyAgent } )
+      dl.updateOptions({ httpsRequestOptions: { agent: proxyAgent } })
     }
 
     dl.on('end', () => console.info('Download Completed'));

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -28,7 +28,6 @@ function download_lib(libname) {
 
     const proxy = process.env.https_proxy ?? process.env.HTTPS_PROXY;
     if (proxy) {
-      console.info("Using configured HTTP proxy: " + proxy)
       const proxyAgent = new HttpsProxyAgent(proxy);
       dl.updateOptions({
           httpsRequestOptions: { agent: proxyAgent },

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -21,7 +21,6 @@ function download_lib(libname) {
 
     const url = `${DOWNLOADS_BASE_URL}/${CURRENT_VERSION}/${libname}`;
     console.info(`Downloading lib ${libname} from ${url}`);
-
     const dl = new DownloaderHelper(url, __dirname, {
         override: true,
     });

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -1,5 +1,6 @@
 const { Agent } = require('https');
 const { DownloaderHelper } = require('node-downloader-helper');
+const { HttpsProxyAgent } = require('https-proxy-agent');
 const { version } = require("./package.json");
 const { platform, arch } = process
 
@@ -21,14 +22,16 @@ function download_lib(libname) {
 
     const url = `${DOWNLOADS_BASE_URL}/${CURRENT_VERSION}/${libname}`;
     console.info(`Downloading lib ${libname} from ${url}`);
+
     const dl = new DownloaderHelper(url, __dirname, {
-        override: true,
-        httpsRequestOptions: {
-          // Disable keepalive to prevent the process hanging open.
-          // https://github.com/matrix-org/matrix-rust-sdk/issues/1160
-          agent: new Agent({ keepAlive: false }),
-        },
+      override: true,
     });
+
+    var proxy = process.env.https_proxy ?? process.env.http_proxy ?? undefined
+    if (proxy) {
+      const proxyAgent = new HttpsProxyAgent(proxy);
+      dl.updateOptions( { agent: proxyAgent } )
+    }
 
     dl.on('end', () => console.info('Download Completed'));
     dl.on('error', (err) => console.info('Download Failed', err));

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -27,10 +27,10 @@ function download_lib(libname) {
 
     const proxy = process.env.https_proxy ?? process.env.HTTPS_PROXY;
     if (proxy) {
-      const proxyAgent = new HttpsProxyAgent(proxy);
-      dl.updateOptions({
-          httpsRequestOptions: { agent: proxyAgent },
-      });
+        const proxyAgent = new HttpsProxyAgent(proxy);
+        dl.updateOptions({
+            httpsRequestOptions: { agent: proxyAgent },
+        });
     }
 
     dl.on('end', () => console.info('Download Completed'));

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -1,4 +1,3 @@
-const { Agent } = require('https');
 const { DownloaderHelper } = require('node-downloader-helper');
 const { HttpsProxyAgent } = require('https-proxy-agent');
 const { version } = require("./package.json");

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -26,7 +26,7 @@ function download_lib(libname) {
       override: true,
     });
 
-    var proxy = process.env.https_proxy ?? process.env.http_proxy ?? undefined
+    var proxy = process.env.https_proxy ?? process.env.HTTPS_PROXY
     if (proxy) {
       console.info("Using configured HTTP proxy: " + proxy)
       const proxyAgent = new HttpsProxyAgent(proxy);

--- a/bindings/matrix-sdk-crypto-nodejs/download-lib.js
+++ b/bindings/matrix-sdk-crypto-nodejs/download-lib.js
@@ -1,5 +1,5 @@
-const { DownloaderHelper } = require('node-downloader-helper');
 const { HttpsProxyAgent } = require('https-proxy-agent');
+const { DownloaderHelper } = require('node-downloader-helper');
 const { version } = require("./package.json");
 const { platform, arch } = process
 
@@ -23,14 +23,16 @@ function download_lib(libname) {
     console.info(`Downloading lib ${libname} from ${url}`);
 
     const dl = new DownloaderHelper(url, __dirname, {
-      override: true,
+        override: true,
     });
 
-    var proxy = process.env.https_proxy ?? process.env.HTTPS_PROXY
+    const proxy = process.env.https_proxy ?? process.env.HTTPS_PROXY;
     if (proxy) {
       console.info("Using configured HTTP proxy: " + proxy)
       const proxyAgent = new HttpsProxyAgent(proxy);
-      dl.updateOptions({ httpsRequestOptions: { agent: proxyAgent } })
+      dl.updateOptions({
+          httpsRequestOptions: { agent: proxyAgent },
+      });
     }
 
     dl.on('end', () => console.info('Download Completed'));

--- a/bindings/matrix-sdk-crypto-nodejs/package.json
+++ b/bindings/matrix-sdk-crypto-nodejs/package.json
@@ -29,6 +29,7 @@
         "doc": "typedoc --tsconfig ."
     },
     "dependencies": {
+        "https-proxy-agent": "^5.0.1",
         "node-downloader-helper": "^2.1.1"
     }
 }

